text/0000-builder-spec.md

```diff
@@ -0,0 +1,295 @@
+		[
+		  {
+			"id": "<buildpack id>",
+			"version": "<buildpack version>",
+			"optional": "<bool>"
+		  }
+		]
+	}
+]
+```
+
+#### Metadata
+The `io.buildpacks.builder.metadata` data should look like:
+```json
+{
+  "description": "<Some description>",
+  "api": "<some builder API version | default 0.1 >",
+  "stack": {
+    "runImage": {
+      "image": "<some/run-image>",
+      "mirrors": [
+        "<gcr.io/some/default>"
+      ]
+    }
+  },
+  "buildpacks": [
+    {
+      "id": "<buildpack id>",
+	  "version": "<buildpack version>",
+	  "homepage": "http://geocities.com/top-bp"
+	}
+  ],
+  "lifecycle": {
+    "version": "<version of lifecycle in builder>",
+    "apis":  {
+        "buildpack": {
+          "deprecated": ["0.1"],
+          "supported": ["1.2", "1.3"]},
+        "platform": {
+          "deprecated": [],
+          "supported": ["2.3", "2.4"]}
+    }
+  },
+  "createdBy": {
+    "name": "<creator of builder>", 
+    "version": "<builder version>"
+  }
+}
+```
+
+#### Layers
+The `io.buildpacks.buildpack.layers` data should look like:
+```json
+{
+  "<buildpack id>": {
+    "<buildpack version>": {
+      "api": "<buildpack API>",
+      "order": [
+        {
+          "group": [
+            {
+              "id": "<inner buildpack>",
+              "version": "<inner bulidpacks version>"
+            }
+          ]
+        }
+      ],
+      "layerDiffID": "sha256:test.nested.sha256",
+	  "homepage": "http://geocities.com/top-bp"
+    }
+  },
+  "<inner buildpack>": {
+    "<inner buildpacks version>": {
+      "api": "<buildpack API>",
+      "stacks": [
+        {
+          "id": "<stack ID buildpack supports>",
+          "mixins": ["<list of mixins required>"]
+        }
+      ],
+      "layerDiffID": "sha256:test.bp.one.sha256",
+	  "homepage": "http://geocities.com/cool-bp"
+    }
+  }
+}
+```
+
+### Lifecycle APIs
+The `io.buildpacks.lifecycle.apis` data should look like:
+```json
+{
+  "buildpack": {
+    "deprecated": ["<list of versions>"],
+    "supported": ["<list of versions>"]
+  },
+  "platform": {
+    "deprecated": ["<list of versions>"],
+    "supported": ["<list of versions>"]
+  }
+}
+```
+
+## Spec Additions
+### Extension Spec
+Builders should be specified as an extension spec, and explicitly versioned, using a `<major>.<minor>` pattern, and following the [established spec api release pattern][spec-api-branches]. The version should be persisted in the `io.buildpacks.builder.metadata`, as a top-level `api` field. Versioning will start with 0.1.
+
+### [Platform specification][platform-spec]
+We should add a cross-reference to the extension builder specification, noting that a builder is a fully-functioning Stack build image.
+
+### [Distribution specification][distribution-spec]
+We should add a cross-reference to the extension builder specification, noting that a builder is also a valid artifact, and is specified in the extension spec.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+N/A
```

Comment:
For the benefits of the formal specification we trade off speed. For example, it will require more ceremony to make changes to builder metadata.